strawberry/scalars.py

```diff
@@ -1,18 +1,35 @@
 from __future__ import annotations
 
 import base64
-from typing import TYPE_CHECKING, Any, Dict, NewType, Union
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    Generic,
+    NewType,
+    Union,
+)
+from typing_extensions import TypeVar
 
 from .custom_scalar import scalar
 
 if TYPE_CHECKING:
     from .custom_scalar import ScalarDefinition, ScalarWrapper
 
-
+T = TypeVar("T", bound=Any, default=Any)
 ID = NewType("ID", str)
 
-JSON = scalar(
-    NewType("JSON", object),  # mypy doesn't like `NewType("name", Any)`
+
+# This is required for mypy/pyright identify JSON[T] as T and not JSON[T]
+# They will respect only the typings on `__getitem__` defined in the metaclass
+# and ignore `__class_getitem__` on both
+class JSONMeta(type):
+    def __getitem__(self, item: T) -> T:
+        return self.__class_getitem__(item)  # type: ignore
```

Comment:
If I understand your problem properly, can you use a Generic Protocol instead?

Reply:
Will try here, makes sense

Reply:
I tried, but using a `Protocol` here, both as a multiple inheritance for `JSONMeta` and also to type `self: GenericProtocol` produced other errors (specially for pyright) and made it stop working.

I tried some other approaches, but couldn't get it to work. Do you have any suggestions that I can try here?